src/stats/kde.nim

```diff
@@ -0,0 +1,209 @@
+  ## If `normalize` is true the result will be normalized such that the
+  ## integral over it is equal to 1.
+  ##
+  ## By default the evaluation points will be `samples` linearly spaced points
+  ## between `[min(t), max(t)]`. If desired the evaluation points can be given
+  ## explicitly by handing a `Tensor[float] | openArray[float]` as `samples`.
+  ##
+  ## The `kernel` is the kernel function that will be used. Unless you want to
+  ## use a custom kernel function, call the convenience wrapper below, which
+  ## only takes a `KernelKind` (either as string or directly as an enum value)
+  ## below, which defaults to a gaussian kernel.
+  ##
+  ## Custom kernel functions are supported by handing a function of signature
+  ##
+  ## `KernelFunc = proc(x, x_i, bw: float): float`
+  ##
+  ## to this procedure and setting the `kernelKind` to `knCustom`. This ``requires``
+  ## to also hand a `cutoff`, which is the window of `s[j] - t[i] <= cutoff`, where
+  ## `s[j]` is the `j`-th sample and `t[i]` the `i`-th input value. Only this window is
+  ## considered for the kernel summation for efficiency. Set it such that the
+  ## contribution of the custom kernel is very small (or 0) outside that range.
+  let N = t.size
+  # `t` not sorted here yet, but does not matter
+  let A = min(std(t),
+              iqr(t) / 1.34)
+  let bwAct = if classify(bw) != fcNaN: bw
+              else: 0.9 * A * pow(N.float, -1.0/5.0)
+
+  var t = t
+  var weights = weights
+  if weights.size > 0:
+    doAssert weights.size == t.size
+    let sortIdx = t.argsort()
+    ## TODO: if we don't clone the resulting t and weights for some reason they end up
+    ## all messed up?!
+    t = t[sortIdx].clone()
+    weights = weights[sortIdx].clone()
+  else:
+    t = t.asType(float).sorted
+  let (minT, maxT) = (min(t), max(t))
+  when U is int:
+    let x = linspace(minT, maxT, samples)
+    let nsamples = samples
+  elif U is seq | array:
+    let x = toTensor(@samples).asType(float)
+    let nsamples = x.size
+  else:
+    let x = samples.asType(float)
+    let nsamples = x.size
+  result = newTensor[float](nsamples)
+  let norm = 1.0 / (N.float * bwAct)
+  var
+    start = 0
+    stop = 0
+  doAssert classify(cutoff) != fcNan or kernelKind != knCustom, "If a custom " &
+    "is used you have to provide a cutoff distance!"
+  let cutoff = if classify(cutoff) != fcNan: cutoff
+               else: getCutoff(bwAct, kernelKind)
+  if weights.size == 0:
+    for i in 0 ..< t.size:
+      (start, stop) = findWindow(cutoff, t[i], x, start, stop)
+      # TODO: rewrite using kernel(t: Tensor) and fancy indexing?
+      for j in start ..< stop:
+        result[j] += norm * kernel(x[j], t[i], bwAct)
+  else:
+    for i in 0 ..< t.size:
+      (start, stop) = findWindow(cutoff, t[i], x, start, stop)
+      # TODO: rewrite using kernel(t: Tensor) and fancy indexing?
+      for j in start ..< stop:
+        result[j] += weights[i] * norm * kernel(x[j], t[i], bwAct)
```

Comment:
I tried to use a template:
```nim
  template loop(arg: untyped): untyped =
    for i {.inject.} in 0 ..< t.size:
      (start, stop) = findWindow(cutoff, t[i], x, start, stop)
      for j {.inject.} in start ..< stop:
        arg
```
but it doesn't work. Fails with undeclared identifier `j`, sigh.

Reply:
What if you use {.dirty.}? Or `mixin j`

I guess the undeclared identifier is due to the template being called in a generic proc

Reply:
Ah, right. Forgot about that "template in generic" gotcha again.

I tried `{.dirty.}`, that didn't help. I'll give `mixin` a try.